activerecord/lib/active_record/associations/through_association.rb

```diff
@@ -71,7 +71,15 @@ def stale_state
         end
 
         def foreign_key_present?
-          through_reflection.belongs_to? && !owner[through_reflection.foreign_key].nil?
+          (through_reflection.belongs_to? && !owner[through_reflection.foreign_key].nil?) ||
+            recursive_through_foreign_key_present?(through_reflection.through_reflection)
```

Comment:
It seems that this should work with the way you suggest, I don't remember why I've wrote the second method. Thanks for the suggestion. I'll have a look at this asap.